keras_nlp/layers/mlm_classification_head.py

```diff
@@ -0,0 +1,204 @@
+
+        self.vocabulary_size = vocabulary_size
+        self.embedding_weights = embedding_weights
+        self.inner_activation = keras.activations.get(inner_activation)
+        self.outer_activation = keras.activations.get(outer_activation)
+        self.layer_norm_epsilon = layer_norm_epsilon
+        self.kernel_initializer = keras.initializers.get(kernel_initializer)
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self._built = False
+
+        if vocabulary_size is None and embedding_weights is None:
+            raise ValueError(
+                "One of `vocabulary_size` or `embedding_weights` must be set. "
+                "Received: `vocabulary_size=None`, `embedding_weights=None`"
+            )
+
+        if embedding_weights is not None:
+            shape = embedding_weights.shape
+            if vocabulary_size is not None and vocabulary_size != shape[0]:
+                raise ValueError(
+                    "`vocabulary_size` should match the first dimension of the "
+                    "shape of `embedding_weights`. Received: "
+                    f"`vocabulary_size={vocabulary_size}`, "
+                    f"`embedding_weights.shape={shape}`"
+                )
+            self.vocabulary_size = shape[0]
+
+    def _build(self, input_shape):
+        # Create layers based on input shape.
+        self._built = True
+        feature_size = input_shape[-1]
+
+        self._dense = keras.layers.Dense(
+            feature_size,
+            activation=self.inner_activation,
+            kernel_initializer=self.kernel_initializer,
+            bias_initializer=self.bias_initializer,
+        )
+        self._layer_norm = tf.keras.layers.LayerNormalization(
+            epsilon=self.layer_norm_epsilon,
+        )
+        if self.embedding_weights is None:
+            self._kernel = self.add_weight(
+                name="output_kernel",
+                shape=[feature_size, self.vocabulary_size],
+                initializer=self.kernel_initializer,
+                dtype=self.dtype,
+            )
+        self._bias = self.add_weight(
+            name="output_bias",
+            shape=[self.vocabulary_size],
+            initializer=self.bias_initializer,
+            dtype=self.dtype,
+        )
+
+    def call(self, inputs, mask_positions):
+        if not self._built:
+            self._build(inputs.shape)
+
+        # Gather the encoded tokens at the masked indices.
+        x = tf.gather(inputs, mask_positions, axis=1, batch_dims=1)
+
+        # Apply a trainable linear transformation and a layer norm.
+        x = self._dense(x)
+        x = self._layer_norm(x)
+
+        # Transform encodings to vocabulary_size predictions.
+        if self.embedding_weights is None:
+            outputs = tf.matmul(x, self._kernel)
+        else:
+            outputs = tf.matmul(
+                x,
+                tf.cast(self.embedding_weights, self.compute_dtype),
```

Comment:
I am not sure about this - if here we do not cast x's dtype, would there be a type mismatch between `x` and `self.embedding_weights`?

Reply:
In mixed precision, x will generally be float16 and embedding_weights float32, so omitting this leads to an error. Not sure if there is a better way to do this though.

Reply:
Yea, I was wondering if we need to cast x's type as well? In the mixed precision case, compute_dtype would just be x's dtype, but would there be some case that x's dtype mismatch `self.compyte_dtype`?

Reply:
Ah I believe the `autocast` functionality on the base Keras layer will handle that?

Reply:
nice to know, thanks!